src/Stratis.Features.SQLiteWalletRepository/IWalletRepository.cs

```diff
@@ -0,0 +1,171 @@
+
+namespace Stratis.Features.SQLiteWalletRepository
+{
+    /// <summary>
+    /// Defines the required interface of a wallet repository.
+    /// </summary>
+    /// <remarks>
+    /// The repository should not contain any business logic other than what is implied or provided by services explicitly associated
+    /// with this interface. These currently include:
+    /// - Network and Consensus. Used for obtaining:
+    ///   - the coin type for HdPath resolution
+    ///   - human-readable addresses to be returned by some public methods
+    /// - DataFolder. Used for obtaining:
+    ///   - the wallet folder.
+    /// - IDateTimeProvider
+    ///   - used for populating CreationTime fields.
+    ///   IScriptAddressReader (or IScriptDestinationReader)
+    ///   - used to find the destinations of <see cref="TxOut.ScriptPubKey" /> scripts.
+    /// </remarks>
+    public interface IWalletRepository
+    {
+        /// <summary>
+        /// Updates the relevant wallets from the information contained in the transaction.
+        /// </summary>
+        /// <param name="walletName">The wallet name.</param>
+        /// <param name="transaction">The transient transaction to process.</param>
+        /// <param name="txId">Used to overrides the default transaction id.</param>
+        /// <remarks>
+        /// This method is intended to be idempotent - i.e. running it twice should not produce any adverse effects.
+        /// If any empty wallet addresses have transactions added to them then the affected accounts should
+        /// have their addresses topped up to ensure there are always a buffer of unused addresses after the last
+        /// address containing transactions.
+        /// </remarks>
+        void ProcessTransaction(string walletName, Transaction transaction, uint256 txId = null);
+
+        /// <summary>
+        /// Updates all the wallets from the information contained in the block.
+        /// </summary>
+        /// <param name="block">The block to process.</param>
+        /// <param name="header">The header of the block passed.</param>
+        /// <param name="walletName">Set this to limit processing to the named wallet.</param>
+        /// <remarks>
+        /// This method is intended to be idempotent - i.e. running it twice consecutively should not produce any adverse effects.
+        /// Similar to the rest of the methods it should not contain any business logic other than what may be injected externally.
+        /// If any empty wallet addresses have transactions added to them then the affected accounts should
+        /// have their addresses topped up to ensure there are always a buffer of unused addresses after the last
+        /// address containing transactions.
+        /// It's the caller's responsibility to ensure that this method is not called again when it's already executing.
+        /// </remarks>
+        void ProcessBlock(Block block, ChainedHeader header, string walletName = null);
```

Comment:
>The repository should not contain any business logic other than what is implied or provided by services explicitly associated

This is already incorrect based on this interface, passing a block to the repo means you will have to use business logic to extract the scripts that the wallet is tracking.

Reply:
> This is already incorrect based on this interface, passing a block to the repo means you will have to use business logic to extract the scripts that the wallet is tracking.

That's what is meant by "other than what is implied or provided by services explicitly associated". The following is listed in the interface summary. As you can see `IScriptAddressReader` is in fact included via DI (and the business logic is defined externally):

```
 /// <summary>
    /// Defines the required interface of a wallet repository.
    /// </summary>
    /// <remarks>
    /// The repository should not contain any business logic other than what is implied or provided by services explicitly associated
    /// with this interface. These currently include:
    /// - Network and Consensus. Used for obtaining:
    ///   - the coin type for HdPath resolution
    ///   - human-readable addresses to be returned by some public methods
    /// - DataFolder. Used for obtaining:
    ///   - the wallet folder.
    /// - IDateTimeProvider
    ///   - used for populating CreationTime fields.
    ///   IScriptAddressReader (or IScriptDestinationReader)
    ///   - used to find the destinations of <see cref="TxOut.ScriptPubKey" /> scripts.
    /// </remarks>
```

Reply:
Then why do you need to mix this in to the Repository, generally repository is a simple insert/read/update class, injecing components to it that hold business logic still to me implies businedd logic, you are effectively making a new WalletManager (so maybe just call it that?).

The repository layer should only take Indexes (sha256) and Scripts (byte arrays) no?

Reply:
You can think of those things along the lines of stored procedure which still falls in the database domain. I.e. the data layer is not necessarily a simple insert/read/update class but it may also contain mechanisms to guarantee its own integrity. The repository is defensive i.t.o. what it allows you to do. It does less than what a WalletManager would so as the latter would contain methods that build on the functionality offered by the repository layer in order to service the controllers.